### PR TITLE
[Integration][Pagerduty] Handle Pagerduty pagination offset

### PR DESCRIPTION
**Pull request by melodyogonna**
### **User description**
# Description

What - Pagerduty rejects requests where `offset` + `page limit` is greater than 10,000. At the moment this crashes the resync, we should instead stop paginating once this limit is bound to be exceeded.

Why - Prevent Pagerduty limit from crashing the integration.

How - Add additional conditon to the while loop.

## Type of change

Please leave one option from the following and delete the rest:

- [ x ] Bug fix (non-breaking change which fixes an issue)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Bug fix

___

### **Description**
- Add pagination limit check to prevent PagerDuty 10k resource limit errors

- Implement early exit when offset + page size exceeds maximum resources

- Add comprehensive test coverage for pagination limit handling

- Clean up minor formatting in analytics method

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["PagerDuty Pagination Loop"] -->|Check offset + PAGE_SIZE| B["Exceeds MAX_PAGERDUTY_RESOURCES?"]
  B -->|Yes| C["Exit Loop Early"]
  B -->|No| D["Continue Fetching Data"]
  D -->|has_more_data| A
  E["Test Coverage"] -->|Validates| C
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pagerduty.py</strong><dd><code>Add pagination limit check to prevent resource overflow</code>&nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/clients/pagerduty.py

<ul><li>Add <code>MAX_PAGERDUTY_RESOURCES</code> constant set to 10,000 with documentation <br>link<br> <li> Update pagination while loop condition to check if <code>(offset + </code><br><code>PAGE_SIZE) < MAX_PAGERDUTY_RESOURCES</code><br> <li> Remove unnecessary blank line in <code>get_incident_analytics_by_services</code> <br>method</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2663/files#diff-79febc4a748111076c6ff542e5813561fd6e51a42ac805e969719e0aa9ebac24">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_client.py</strong><dd><code>Add test for pagination limit boundary condition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/pagerduty/tests/test_client.py

<ul><li>Add new test <code>test_paginate_request_to_pager_duty_max_limit</code> to verify <br>early exit behavior<br> <li> Test creates mock responses for pages up to the limit minus one<br> <li> Validates that pagination stops before exceeding the 10k resource <br>threshold<br> <li> Add type hint <code>list[Any]</code> for mock_responses to satisfy mypy</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2663/files#diff-9909f1f9ed695bc8b6cf7334048069b8073f645654a3d8cf3f9474cbc61e0192">+27/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/pagerduty/clients/pagerduty.py [hunk 1]
```diff
@@ -94,6 +96,14 @@ async def paginate_request_to_pager_duty(
...
                     offset += data["limit"]
             except httpx.HTTPStatusError as e:
+                if (
+                    e.response.status_code == 400
+                    and (offset + PAGE_SIZE) >= MAX_PAGERDUTY_RESOURCES
+                ):
+                    logger.warning(
+                        f"Reached max pagerduty limit of {MAX_PAGERDUTY_RESOURCES} for {resource}"
+                    )
+                    break
```
Comment (mk-armah) [inline integrations/pagerduty/clients/pagerduty.py]: This might never be hit

File: integrations/pagerduty/clients/pagerduty.py [hunk 2]
```diff
@@ -75,7 +77,7 @@ async def paginate_request_to_pager_duty(
         offset = 0
         has_more_data = True
 
-        while has_more_data:
+        while has_more_data and offset < MAX_PAGERDUTY_RESOURCES:
```
Comment (mk-armah) [inline integrations/pagerduty/clients/pagerduty.py]: if we create a exit condition 
` offset < MAX_PAGERDUTY_RESOURCES:`
within the while loop itself, we can use that to exit when `MAX_PAGERDUTY_RESOURCES` is reached and also take advantage of the warning log effectively
